runtime/common/src/evm/precompile.rs

```diff
@@ -12,129 +12,70 @@
-			BLAKE2F_ADDR => Some(Blake2F::execute(handle)),
-			SHA3FIPS256_ADDR => Some(Sha3FIPS256::execute(handle)),
-			DISPATCH_ADDR => Some(Dispatch::<R>::execute(handle)),
-			ECRECOVERPUBLICKEY_ADDR => Some(ECRecoverPublicKey::execute(handle)),
-			LP_AXELAR_GATEWAY => {
-				Some(<axelar_gateway_precompile::Pallet<R> as Precompile>::execute(handle))
-			}
-			_ => None,
-		}
+	fn decimals() -> u8 {
+		18
 	}
 
-	fn is_precompile(&self, address: H160, _remaining_gas: u64) -> IsPrecompileResult {
-		IsPrecompileResult::Answer {
-			is_precompile: matches!(
-				address.0,
-				ECRECOVER_ADDR
-					| SHA256_ADDR | RIPEMD160_ADDR
-					| IDENTITY_ADDR | MODEXP_ADDR
-					| BN128ADD_ADDR | BN128MUL_ADDR
-					| BN128PAIRING_ADDR | BLAKE2F_ADDR
-					| SHA3FIPS256_ADDR | DISPATCH_ADDR
-					| ECRECOVERPUBLICKEY_ADDR
-					| LP_AXELAR_GATEWAY
-			),
-			extra_cost: 0,
-		}
+	fn is_native_currency() -> bool {
+		true
 	}
 }
 
-/// Altair's precompiles
-/// For now, Altair uses the exact same set of precompiles used in Development.
-pub type Altair<R> = CentrifugePrecompiles<R>;
+type EthereumPrecompilesChecks = (AcceptDelegateCall, CallableByContract, CallableByPrecompile);
 
-/// A set of precompiles. This set might contain
-/// not yet mainnet ready precompiles in order to test
-/// those in development or staging environment without touching
-/// the mainnet set.
-pub type Development<R> = CentrifugePrecompiles<R>;
+// Address numbers comes from:
+// - https://docs.moonbeam.network/builders/pallets-precompiles/precompiles/overview/#precompiled-contract-addresses
+// - https://github.com/centrifuge/liquidity-pools/blob/release-v1.0/src/gateway/routers/axelar/Forwarder.sol#L29
 
-// H160 cannot be used in a match statement due to its hand-rolled
-// PartialEq implementation. This just gives a nice name to the
-// internal array of bytes that an H160 wraps.
-type Addr = [u8; 20];
+#[precompile_utils::precompile_name_from_address]
+pub type RuntimePrecompilesAt<R, Symbol> = (
+	// Ethereum precompiles:
+	// We allow DELEGATECALL to stay compliant with Ethereum behavior.
+	PrecompileAt<AddressU64<0x1>, ECRecover, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x2>, Sha256, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x3>, Ripemd160, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x4>, Identity, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x5>, Modexp, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x6>, Bn128Add, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x7>, Bn128Mul, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x8>, Bn128Pairing, EthereumPrecompilesChecks>,
+	PrecompileAt<AddressU64<0x9>, Blake2F, EthereumPrecompilesChecks>,
+	// Non-Moonbeam specific nor Ethereum precompiles:
+	PrecompileAt<AddressU64<0x400>, Sha3FIPS256, (CallableByContract, CallableByPrecompile)>,
+	PrecompileAt<AddressU64<0x401>, Dispatch<R>, (CallableByContract, CallableByPrecompile)>,
```

Comment:
This is disabled on moonbeam. Lets not make it callable by anything but a user!

Reply:
Yes, it is disabled, but we had it enabled, that's why I left so.

> Lets not make it callable by anything but a user!

I'm not sure what this means. Should I replicate the same as Moonbeam here for this case?

Reply:
Moonbean has this as: https://github.com/moonbeam-foundation/moonbeam/blob/a96e21261750055cd19c0a2019fd66cdd7090814/runtime/moonbeam/src/precompiles.rs#L108
- Option 1: as Moonbeam,
- Option 2: `PrecompileAt<AddressU64<0x401>, Dispatch<R>>` (without any check)

Reply:
I would want it to be able to only be called by user, not by contract, not by precompile, not by delegate call.

Reply:
Then I think it should be option 2